README.md
# Final-Physics-Lab-Code

Codes for the final physics lab I have created and pieced to pieced together.


changelog
Iteration 1: Simple pendulum using the Verlet method.

Iteration 2: Coupled pendulum with spring connection.

Iteration 3: Coupled pendulum with shaking board connection.

Iteration 4: Position on the X-axis against time with the kuramoto model (give length of pendulum)

Iteration 5: Angle against time with the kuramoto model (give length of pendulum)

Iteration 6: Position on the X-axis against time with the kuramoto model (give period of pendulum) (using real values)

Iteration 7: Angle against time with the kuramoto model (give period of pendulum) (using real values)

Iteration 8: Coupled spring-mass system


Iteration 1.py
import numpy as np
import matplotlib.pyplot as plt

# https://medium.com/modern-physics/simple-pendulum-odesolver-using-python-dcb30c267eee

class ODESolver(object):
    """Second-order ODE Solver.
    Parameters
    ------------
    omega_0 : float
        initial angular velocity
    theta_0 : float
        initial angular displacement
    eta : float
        time step size
    n_iter : int
        number of steps
        
    Attributes
    -----------
    time_ : 1d-array
        Stores time values for each time step.
    omega_ : 1d-array
        Stores angular velocity values for each time step.
    theta_ : 1d-arra
        Stores angular displacement values for each time step.
        
    Methods
    -----------
    euler(alpha): Implements the Euler algorithm for the acceleration function alpha.
    
    midpoint(alpha): Implements the Midpoint algorithm for the acceleration function alpha.
    
    verlet(alpha): Implements the Verlet algorithm for the acceleration function alpha.
    """
    def __init__(self, omega_0 = 0, theta_0 = 10, eta=0.01, n_iter=10):
        self.omega_0 = omega_0
        self.theta_0 = theta_0
        self.eta = eta
        self.n_iter = n_iter
        
    
    
    def verlet(self,alpha):
        """Implement Verlet Method.
        
        Parameters
        ----------
        alpha : acceleration function
        Returns
        -------
        self : object
        """
        self.time_ = np.zeros(self.n_iter)
        self.theta_ = np.zeros(self.n_iter)
        self.alpha_ = np.zeros(self.n_iter)
        self.theta_[0] = self.theta_0*np.pi/180.0
        self.alpha_[0] = alpha(self.theta_[0])
        self.time_[1]= self.eta
        self.theta_[1] = self.theta_[0]+self.omega_0*self.eta +0.5* (self.eta**2)*alpha(self.theta_[0])
        self.alpha_[1] = alpha(self.theta_[1])
        
        
        for i in range(self.n_iter-2):
            self.time_[i+2] = self.time_[i+1] + self.eta
            self.theta_[i+2] = 2.0*self.theta_[i+1] -self.theta_[i] + (self.eta**2)*alpha(self.theta_[i+1])
            self.alpha_[i+2] = alpha(self.theta_[i+2])
        return self

def alpha(x):
    g = 9.80665 # m/s^2
    l = 10 # m
    return -g / l * np.sin(x)

pendulum = ODESolver(omega_0 = 0, theta_0 = 10, eta=0.1, n_iter=300).verlet(alpha)
time = pendulum.time_
theta = pendulum.theta_
alphas = pendulum.alpha_
plt.plot(time,theta*180/np.pi,lw=3,color='blue', label = 'theta')
#plt.plot(time,alphas*180/np.pi,lw=3,color='red', label = 'acceleration')
plt.legend()
plt.xlabel('time(s)',size=13)
plt.ylabel('angle (deg)',size=13)
plt.title('Verlet Method',size=13)
plt.show()

Iteration 3.py
import numpy as np
import matplotlib.pyplot as plt

# https://medium.com/modern-physics/simple-pendulum-odesolver-using-python-dcb30c267eee

class ODESolver(object):
    """Second-order ODE Solver.
    Parameters
    ------------
    omega_0 : float
        initial angular velocity
    theta_0 : float
        initial angular displacement of the right pendulum
    phi_0 : float
        initial angular displacement of the left pendulum
    eta : float
        time step size
    n_iter : int
        number of steps
    theta_mass : float
        mass of the right pendulum's bob
    phi_mass : float
        mass of the left pendulum's bob
    system_mass : float
        mass of the the system excluding the bobs

        
    Attributes
    -----------
    time_ : 1d-array
        Stores time values for each time step.
    omega_ : 1d-array
        Stores angular velocity values for each time step.
    theta_ : 1d-arra
       Stores angular displacement values for each time step.
        
    Methods
    -----------
    verlet(alpha): Implements the Verlet algorithm for the acceleration function alpha.
    """
    def __init__(self, omega_0, theta_0, phi_0, eta, n_iter, theta_mass, phi_mass):
        self.omega_0 = omega_0
        self.theta_0 = theta_0
        self.phi_0 = phi_0
        self.eta = eta
        self.n_iter = n_iter
        self.theta_mass = theta_mass
        self.phi_mass = phi_mass
        
    
    
    def verlet(self,alpha):
        """Implement Verlet Method.
        
        Parameters
        ----------
        alpha : acceleration function
        Returns
        -------
        self : object
        """
        self.time_ = np.zeros(self.n_iter)
        self.theta_ = np.zeros(self.n_iter)
        self.phi_ = np.zeros(self.n_iter)
        self.theta_[0] = self.theta_0*np.pi/180.0
        self.phi_[0] = self.phi_0*np.pi/180.0
        self.time_[1]= self.eta
        self.theta_[1] = self.theta_[0]+self.omega_0*self.eta +0.5* (self.eta**2)*alpha(self.theta_[0], self.phi_[0], self.theta_mass, self.phi_mass)
        self.phi_[1] = self.phi_[0]+self.omega_0*self.eta +0.5* (self.eta**2)*alpha(self.phi_[0], self.theta_[0], self.phi_mass, self.theta_mass)
        
        for i in range(self.n_iter-2):
            self.time_[i+2] = self.time_[i+1] + self.eta
            self.theta_[i+2] = 2.0*self.theta_[i+1] -self.theta_[i] + (self.eta**2)*alpha(self.theta_[i+1], self.phi_[i+1], self.theta_mass, self.phi_mass)
            self.phi_[i+2] = 2.0*self.phi_[i+1] -self.phi_[i] + (self.eta**2)*alpha(self.phi_[i+1], self.theta_[i+1], self.phi_mass, self.theta_mass)
        return self

def alpha(x, y, mx, my):
    g = 9.80665 # m/s^2
    l = 10 # m
    asq = (-g/l*np.sin(x) * np.sin(x) + -g/l*np.sin(y) * np.cos(y) * mx / my)**2 + -g/l*np.sin(x) * np.cos(x)
    if asq > 0:
        a = asq**0.5
    else:
        a = -(-asq)**0.5
    print(x)
    return a

pendulum = ODESolver(omega_0 = 0, theta_0 = 10, phi_0 = 0, eta=0.1, n_iter=100, theta_mass = 1, phi_mass = 1).verlet(alpha)
time = pendulum.time_
theta = pendulum.theta_
phi = pendulum.phi_
plt.plot(time,theta*180/np.pi,lw=3,color='blue', label = 'theta')
plt.plot(time,phi*180/np.pi,lw=3,color='red', label = 'phi')
plt.legend()
plt.xlabel('time(s)',size=13)
plt.ylabel('angle (deg)',size=13)
plt.title('Verlet Method',size=13)
plt.show()

Iteration 4.py
import numpy as np
import matplotlib.pyplot as plt


# https://medium.com/modern-physics/simple-pendulum-odesolver-using-python-dcb30c267eee

def equation(x, y, length, k):
    g = 9.80665  # m/s^2
    return (g / length) ** 0.5 * 2 * np.pi + k * np.sin(y - x)


class Pendulum(object):
    """
    Parameters
    ------------
    theta_0 : float
        initial angular displacement of the right pendulum
    phi_0 : float
        initial angular displacement of the left pendulum
    eta : float
        time step size
    n_iter : int
        number of steps
    theta_length:
        length of the right pendulum
    phi_length:
        length of the left pendulum
    coupling_constant:
        the coupling strength of the pendulum

        
    Attributes
    -----------
    times : 1d-array
        Stores time values for each time step.
    thetas : 1d-array
       Stores angular displacement values for each time step of the right pendulum.
    phis : 1d-array
       Stores angular displacement values for each time step of the left pendulum.
        
    Methods
    -----------
    kuramoto_model:
        Runs the Kuramoto model.
    kuramoto:
        The equation of the Kuramoto model.
    """

    def __init__(self, theta_0, phi_0, eta, n_iter, theta_length, phi_length, coupling_constant):
        self.theta_0 = theta_0
        self.phi_0 = phi_0
        self.eta = eta
        self.n_iter = n_iter
        self.theta_length = theta_length
        self.phi_length = phi_length
        self.coupling_constant = coupling_constant
        self.times = np.zeros(self.n_iter)
        self.thetas = np.zeros(self.n_iter)
        self.phis = np.zeros(self.n_iter)

    def kuramoto_model(self):
        self.thetas[0] = self.theta_0 * np.pi / 180.0
        self.phis[0] = self.phi_0 * np.pi / 180.0

        for i in range(self.n_iter - 1):
            self.times[i + 1] = self.times[i] + self.eta
            self.thetas[i + 1] = self.thetas[i] + self.eta * equation(self.thetas[i], self.phis[i], self.theta_length,
                                                                      self.coupling_constant)
            # print(self.equation(self.thetas[i], self.phis[i], self.theta_length, self.coupling_constant))
            self.phis[i + 1] = self.phis[i] + self.eta * equation(self.phis[i], self.thetas[i], self.phi_length,
                                                                  self.coupling_constant)
            # print(self.equation(self.phis[i], self.thetas[i], self.phi_length, self.coupling_constant))
        return self


system = Pendulum(theta_0=-90, phi_0=0, eta=0.01, n_iter=20, theta_length=0.127, phi_length=0.127,
                  coupling_constant=1).kuramoto_model()
time = system.times * 180 / np.pi
right_pendulum = np.sin(system.thetas) * 0.127
left_pendulum = np.sin(system.phis) * 0.127

plt.plot(time, right_pendulum, lw=3, color='blue', label='theta')
plt.plot(time, left_pendulum, lw=3, color='red', label='phi')
plt.legend()
plt.xlabel('Time (s)', size=13)
plt.ylabel('X-Position (m)', size=13)
plt.title('X-Position Against Time', size=13)
plt.show()


Iteration 5.py
import numpy as np
import matplotlib.pyplot as plt


# https://medium.com/modern-physics/simple-pendulum-odesolver-using-python-dcb30c267eee

def equation(x, y, length, k):
    g = 9.80665  # m/s^2
    return (2 * g / length) ** 0.5 + k * np.sin(y - x)


class Pendulum(object):
    """
    Parameters
    ------------
    theta_0 : float
        initial angular displacement of the right pendulum
    phi_0 : float
        initial angular displacement of the left pendulum
    eta : float
        time step size
    n_iter : int
        number of steps
    theta_length:
        length of the right pendulum
    phi_length:
        length of the left pendulum
    coupling_constant:
        the coupling strength of the pendulum

        
    Attributes
    -----------
    times : 1d-array
        Stores time values for each time step.
    thetas : 1d-array
       Stores angular displacement values for each time step of the right pendulum.
    phis : 1d-array
       Stores angular displacement values for each time step of the left pendulum.
        
    Methods
    -----------
    kuramoto_model:
        Runs the Kuramoto model.
    kuramoto:
        The equation of the Kuramoto model.
    """

    def __init__(self, theta_0, phi_0, eta, n_iter, theta_length, phi_length, coupling_constant):
        self.theta_0 = theta_0
        self.phi_0 = phi_0
        self.eta = eta
        self.n_iter = n_iter
        self.theta_length = theta_length
        self.phi_length = phi_length
        self.coupling_constant = coupling_constant
        self.times = np.zeros(self.n_iter)
        self.thetas = np.zeros(self.n_iter)
        self.phis = np.zeros(self.n_iter)

    def kuramoto_model(self):
        self.thetas[0] = self.theta_0 * np.pi / 180.0
        self.phis[0] = self.phi_0 * np.pi / 180.0

        for i in range(self.n_iter - 1):
            self.times[i + 1] = self.times[i] + self.eta
            self.thetas[i + 1] = self.thetas[i] + self.eta * equation(self.thetas[i], self.phis[i], self.theta_length,
                                                                      self.coupling_constant)
            # print(self.equation(self.thetas[i], self.phis[i], self.theta_length, self.coupling_constant))
            self.phis[i + 1] = self.phis[i] + self.eta * equation(self.phis[i], self.thetas[i], self.phi_length,
                                                                  self.coupling_constant)
            # print(self.equation(self.phis[i], self.thetas[i], self.phi_length, self.coupling_constant))
        return self


system = Pendulum(theta_0=90, phi_0=30, eta=0.1, n_iter=100, theta_length=0.127, phi_length=0.127,
                  coupling_constant=0.5).kuramoto_model()
time = system.times * 180 / np.pi
right_pendulum = np.arcsin(np.sin(system.thetas)) * 180/np.pi
left_pendulum = np.arcsin(np.sin(system.phis)) * 180/np.pi

plt.plot(time, right_pendulum, lw=3, color='blue', label='theta')
plt.plot(time, left_pendulum, lw=3, color='red', label='phi')
plt.legend()
plt.xlabel('Time (s)', size=13)
plt.ylabel('Angle (deg)', size=13)
plt.title('Angle Against Time', size=13)
plt.show()


Iteration 6.py
import numpy as np
import matplotlib.pyplot as plt


# https://medium.com/modern-physics/simple-pendulum-odesolver-using-python-dcb30c267eee

def equation(x, y, f, k):
    return f ** 0.5 * 2 * np.pi + k * np.sin(y - x)


class Pendulum(object):
    """
    Parameters
    ------------
    theta_0 : float
        initial angular displacement of the left pendulum
    phi_0 : float
        initial angular displacement of the right pendulum
    eta : float
        time step size
    n_iter : int
        number of steps
    frequency:
        frequency of both pendulums
    coupling_constant:
        the coupling strength of the pendulum

        
    Attributes
    -----------
    times : 1d-array
        Stores time values for each time step.
    thetas : 1d-array
       Stores angular displacement values for each time step of the left pendulum.
    phis : 1d-array
       Stores angular displacement values for each time step of the right pendulum.
        
    Methods
    -----------
    kuramoto_model:
        Runs the Kuramoto model.
    kuramoto:
        The equation of the Kuramoto model.
    """

    def __init__(self, theta_0, phi_0, eta, n_iter, frequency, coupling_constant):
        self.theta_0 = theta_0
        self.phi_0 = phi_0
        self.eta = eta
        self.n_iter = n_iter
        self.frequency = frequency
        self.coupling_constant = coupling_constant
        self.times = np.zeros(self.n_iter)
        self.thetas = np.zeros(self.n_iter)
        self.phis = np.zeros(self.n_iter)

    def kuramoto_model(self):
        self.thetas[0] = self.theta_0 * np.pi / 180.0
        self.phis[0] = self.phi_0 * np.pi / 180.0

        for i in range(self.n_iter - 1):
            self.times[i + 1] = self.times[i] + self.eta
            self.thetas[i + 1] = self.thetas[i] + self.eta * equation(self.thetas[i], self.phis[i], self.frequency,
                                                                      self.coupling_constant)
            # print(self.equation(self.thetas[i], self.phis[i], self.theta_length, self.coupling_constant))
            self.phis[i + 1] = self.phis[i] + self.eta * equation(self.phis[i], self.thetas[i], self.frequency,
                                                                  self.coupling_constant)
            # print(self.equation(self.phis[i], self.thetas[i], self.phi_length, self.coupling_constant))
        return self


system = Pendulum(theta_0=-90, phi_0=89, eta=0.001, n_iter=700, frequency=184*60/2, coupling_constant=6.5).kuramoto_model()
time = system.times * 180 / np.pi
left_pendulum = np.sin(system.thetas) * 0.127
right_pendulum = np.sin(system.phis) * 0.127

plt.figure(figsize=(14, 6.4))
plt.plot(time, left_pendulum, lw=3, color='blue', label='Left pendulum')
plt.plot(time, right_pendulum, lw=3, color='red', label='Right pendulum')
plt.legend()
plt.xlabel('Time (s)', size=13)
plt.ylabel('X-Position (m)', size=13)
plt.title('X-Position Against Time', size=13)
plt.show()


Iteration 7.py
import numpy as np
import matplotlib.pyplot as plt


# https://medium.com/modern-physics/simple-pendulum-odesolver-using-python-dcb30c267eee

def equation(x, y, f, k):
    return f ** 0.5 * 2 * np.pi + k * np.sin(y - x)


class Pendulum(object):
    """
    Parameters
    ------------
    theta_0 : float
        initial angular displacement of the left pendulum
    phi_0 : float
        initial angular displacement of the right pendulum
    eta : float
        time step size
    n_iter : int
        number of steps
    frequency:
        frequency of both pendulums
    coupling_constant:
        the coupling strength of the pendulum

        
    Attributes
    -----------
    times : 1d-array
        Stores time values for each time step.
    thetas : 1d-array
       Stores angular displacement values for each time step of the left pendulum.
    phis : 1d-array
       Stores angular displacement values for each time step of the right pendulum.
        
    Methods
    -----------
    kuramoto_model:
        Runs the Kuramoto model.
    kuramoto:
        The equation of the Kuramoto model.
    """

    def __init__(self, theta_0, phi_0, eta, n_iter, frequency, coupling_constant):
        self.theta_0 = theta_0
        self.phi_0 = phi_0
        self.eta = eta
        self.n_iter = n_iter
        self.frequency = frequency
        self.coupling_constant = coupling_constant
        self.times = np.zeros(self.n_iter)
        self.thetas = np.zeros(self.n_iter)
        self.phis = np.zeros(self.n_iter)

    def kuramoto_model(self):
        self.thetas[0] = self.theta_0 * np.pi / 180.0
        self.phis[0] = self.phi_0 * np.pi / 180.0

        for i in range(self.n_iter - 1):
            self.times[i + 1] = self.times[i] + self.eta
            self.thetas[i + 1] = self.thetas[i] + self.eta * equation(self.thetas[i], self.phis[i], self.frequency,
                                                                      self.coupling_constant)
            # print(self.equation(self.thetas[i], self.phis[i], self.theta_length, self.coupling_constant))
            self.phis[i + 1] = self.phis[i] + self.eta * equation(self.phis[i], self.thetas[i], self.frequency,
                                                                  self.coupling_constant)
            # print(self.equation(self.phis[i], self.thetas[i], self.phi_length, self.coupling_constant))
        return self


system = Pendulum(theta_0=-90, phi_0=89, eta=0.001, n_iter=700, frequency=184*60/2, coupling_constant=6.5).kuramoto_model()
time = system.times * 180 / np.pi
left_pendulum = np.arcsin(np.sin(system.thetas)) * 180/np.pi
right_pendulum = np.arcsin(np.sin(system.phis)) * 180/np.pi

plt.figure(figsize=(14, 6.4))
plt.plot(time, left_pendulum, lw=3, color='blue', label='Left pendulum')
plt.plot(time, right_pendulum, lw=3, color='red', label='Right pendulum')
plt.legend()
plt.xlabel('Time (s)', size=13)
plt.ylabel('Angle (deg)', size=13)
plt.title('Angle Against Time', size=13)
plt.show()


iteration 8.py
# source: https://scipy-cookbook.readthedocs.io/items/CoupledSpringMassSystem.html


def vectorfield(w, t, p):
    """
    Defines the differential equations for the coupled spring-mass system.

    Arguments:
        w :  vector of the state variables:
                  w = [x1,y1,x2,y2]
        t :  time
        p :  vector of the parameters:
                  p = [m1,m2,k1,k2,L1,L2,b1,b2]
    """
    x1, y1, x2, y2 = w
    m1, m2, k1, k2, L1, L2, b1, b2 = p

    # Create f = (x1',y1',x2',y2'):
    f = [y1,
         (-b1 * y1 - k1 * (x1 - L1) + k2 * (x2 - x1 - L2)) / m1,
         y2,
         (-b2 * y2 - k2 * (x2 - x1 - L2)) / m2]
    return f


# Use ODEINT to solve the differential equations defined by the vector field
from scipy.integrate import odeint

# Parameter values
# Masses:
m1 = 1.0
m2 = 1.5
# Spring constants
k1 = 8.0
k2 = 40.0
# Natural lengths
L1 = 0.5
L2 = 1.0
# Friction coefficients
b1 = 0.8
b2 = 0.5

# Initial conditions
# x1 and x2 are the initial displacements; y1 and y2 are the initial velocities
x1 = 0.5
y1 = 0.0
x2 = 2.25
y2 = 0.0

# ODE solver parameters
abserr = 1.0e-8
relerr = 1.0e-6
stoptime = 10.0
numpoints = 250

# Create the time samples for the output of the ODE solver.
# I use a large number of points, only because I want to make
# a plot of the solution that looks nice.
t = [stoptime * float(i) / (numpoints - 1) for i in range(numpoints)]

# Pack up the parameters and initial conditions:
p = [m1, m2, k1, k2, L1, L2, b1, b2]
w0 = [x1, y1, x2, y2]

# Call the ODE solver.
wsol = odeint(vectorfield, w0, t, args=(p,), atol=abserr, rtol=relerr)

t_ = []
x1_ = []
x2_ = []

for t1, w1 in zip(t, wsol):
    t_.append(t1)
    x1_.append(w1[0])
    x2_.append(w1[2])
# Plot the solution that was generated

print(x1_, x2_)

import matplotlib.pyplot as plt

plt.figure(1, figsize=(14, 6.4))

plt.xlabel('Time (s)')
plt.ylabel('X-Position (m)')
plt.grid(True)

plt.plot(t_, x1_, 'b', linewidth=1, label='$x_1$')
plt.plot(t_, x2_, 'g', linewidth=1, label='$x_2$')

plt.legend()
plt.title('Mass Displacements for the\nCoupled Spring-Mass System')
plt.show()


Iteration 2.py
import numpy as np
import matplotlib.pyplot as plt

# https://medium.com/modern-physics/simple-pendulum-odesolver-using-python-dcb30c267eee

class ODESolver(object):
    """Second-order ODE Solver.
    Parameters
    ------------
    omega_0 : float
            initial angular velocity
    theta_0 : float
            initial angular displacement
    eta : float
        time step size
    n_iter : int
           number of steps
        
    Attributes
    -----------
    time_ : 1d-array
        Stores time values for each time step.
    omega_ : 1d-array
        Stores angular velocity values for each time step.
    theta_ : 1d-arra
       Stores angular displacement values for each time step.
        
    Methods
    -----------
    euler(alpha): Implements the Euler algorithm for the acceleration function alpha.
    
    midpoint(alpha): Implements the Midpoint algorithm for the acceleration function alpha.
    
    verlet(alpha): Implements the Verlet algorithm for the acceleration function alpha.
    """
    def __init__(self, omega_0, theta_0, phi_0, eta, n_iter, theta_mass, phi_mass, spring_distance_theta, spring_distance_phi, spring_constant, spring_length):
        self.omega_0 = omega_0
        self.theta_0 = theta_0
        self.phi_0 = phi_0
        self.eta = eta
        self.n_iter = n_iter
        self.theta_mass = theta_mass
        self.phi_mass = phi_mass
        self.spring_distance_theta = spring_distance_theta
        self.spring_distance_phi = spring_distance_phi
        self.spring_constant = spring_constant
        self.spring_length = spring_length
        
    
    
    def verlet(self,alpha):
        """Implement Verlet Method.
        
        Parameters
        ----------
        alpha : acceleration function
        Returns
        -------
        self : object
        """
        self.time_ = np.zeros(self.n_iter)
        self.theta_ = np.zeros(self.n_iter)
        self.phi_ = np.zeros(self.n_iter)
        self.theta_[0] = self.theta_0*np.pi/180.0
        self.phi_[0] = self.phi_0*np.pi/180.0
        self.time_[1]= self.eta
        self.theta_[1] = self.theta_[0]+self.omega_0*self.eta +0.5* (self.eta**2)*alpha(self.theta_[0], self.phi_[0], self.spring_length, self.spring_constant, self.theta_mass)
        self.phi_[1] = self.phi_[0]+self.omega_0*self.eta +0.5* (self.eta**2)*alpha(self.phi_[0], self.theta_[0], self.spring_length, self.spring_constant, self.phi_mass)
        
        for i in range(self.n_iter-2):
            self.time_[i+2] = self.time_[i+1] + self.eta
            self.theta_[i+2] = 2.0*self.theta_[i+1] -self.theta_[i] + (self.eta**2)*alpha(self.theta_[i+1], self.phi_[i+1], self.spring_length, self.spring_constant, self.theta_mass)
            self.phi_[i+2] = 2.0*self.phi_[i+1] -self.phi_[i] + (self.eta**2)*alpha(self.phi_[i+1], self.theta_[i+1], self.spring_length, self.spring_constant, self.phi_mass)
        return self

def alpha(x, y, sl, sc, m):
    g = 9.80665 # m/s^2
    l = 10 # m
    dis = ((np.cos(x) - np.cos(y))**2 + (np.sin(x) - np.sin(y) + sl)**2)**0.5
    return -g / l * np.sin(x) - (dis - sl)*sc/m * np.sin((np.sin(x) - np.sin(y) + sl)/dis)

pendulum = ODESolver(omega_0 = 0, theta_0 = 10, phi_0 = 0, eta=0.1, n_iter=100, theta_mass = 1, phi_mass = 1, spring_distance_theta = 5, spring_distance_phi = 5, spring_constant = 0.5, spring_length = 2).verlet(alpha)
time = pendulum.time_
theta = pendulum.theta_
phi = pendulum.phi_
plt.plot(time,theta*180/np.pi,lw=3,color='blue', label = 'theta')
plt.plot(time,phi*180/np.pi,lw=3,color='red', label = 'phi')
plt.legend()
plt.xlabel('time(s)',size=13)
plt.ylabel('angle (deg)',size=13)
plt.title('Verlet Method',size=13)
plt.show()